File: write_everyday/1144.递减元素使数组呈锯齿状.go
/*
给你一个整数数组 nums，每次 操作 会从中选择一个元素并 将该元素的值减少 1。

如果符合下列情况之一，则数组 A 就是 锯齿数组：
*/
package main

import "math"

func movesToMakeZigzag(nums []int) int {
	s := [2]int{}
	for i, x := range nums {
		left, right := math.MaxInt, math.MaxInt
		if i > 0 {
			left = nums[i-1]
		}
		if i < len(nums)-1 {
			right = nums[i+1]
		}
		s[i%2] += max(x-min(left, right)+1, 0)
	}
	return min(s[0], s[1])
}

func min(a, b int) int {
	if a > b {
		return b
	}
	return a
}
func max(a, b int) int {
	if a < b {
		return b
	}
	return a
}

File: write_everyday/1144.递减元素使数组呈锯齿状.py
"""
给你一个整数数组 nums，每次 操作 会从中选择一个元素并 将该元素的值减少 1。

如果符合下列情况之一，则数组 A 就是 锯齿数组：
"""
def movesToMakeZigzag(self, nums: list[int]) -> int:
    ret = [0] * 2
    for index, num in enumerate(nums):
        left = nums[index-1] if index > 0 else float('inf')
        right = nums[index+1] if index < len(nums)-1 else float('inf')
        ret[index % 2] += max(0, num - min(left, right) + 1)
    return min(ret)
